lib/src/device_authorization_grant.dart

```diff
@@ -0,0 +1,401 @@
+
+    var headers = <String, String>{};
+    var body = {
+      'device_code': _deviceCode,
+      'grant_type': 'urn:ietf:params:oauth:grant-type:device_code',
+    };
+
+    var secret = this.secret;
+    if (_basicAuth && secret != null && useSecret) {
+      headers['Authorization'] = basicAuthHeader(identifier, secret);
+    } else {
+      // The ID is required for this request any time basic auth isn't being
+      // used, even if there's no actual client authentication to be done.
+      body['client_id'] = identifier;
+      if (secret != null && useSecret) body['client_secret'] = secret;
+    }
+
+    var response =
+        await _httpClient.post(tokenEndpoint, headers: headers, body: body);
+
+    var credentials = handleAccessTokenResponse(
+        response, tokenEndpoint, startTime, _scopes, _delimiter,
+        getParameters: _getParameters);
+
+    _state = _State.finished;
+
+    return Client(
+      credentials,
+      identifier: identifier,
+      secret: secret,
+      basicAuth: _basicAuth,
+      httpClient: _httpClient,
+      onCredentialsRefreshed: _onCredentialsRefreshed,
+    );
+  }
+
+  DeviceProperties _handleDeviceCodeResponse(
+      http.Response response, DateTime startTime, List<String>? scopes,
+      {Map<String, dynamic> Function(MediaType? contentType, String body)?
+          getParameters}) {
+    getParameters ??= parseJsonParameters;
+
+    try {
+      if (response.statusCode != 200) {
+        if (response.statusCode != 400 && response.statusCode != 401) {
+          var reason = '';
+          var reasonPhrase = response.reasonPhrase;
+          if (reasonPhrase != null && reasonPhrase.isNotEmpty) {
+            reason = ' $reasonPhrase';
+          }
+          throw FormatException('OAuth request for "$deviceEndpoint" failed '
+              'with status ${response.statusCode}$reason.\n\n${response.body}');
+        }
+
+        var contentTypeString = response.headers['content-type'];
+        var contentType = contentTypeString == null
+            ? null
+            : MediaType.parse(contentTypeString);
+
+        var parameters = getParameters(contentType, response.body);
+
+        if (!parameters.containsKey('error')) {
+          throw FormatException('did not contain required parameter "error"');
+        } else if (parameters['error'] is! String) {
+          throw FormatException(
+              'required parameter "error" was not a string, was '
+              '"${parameters["error"]}"');
+        }
+
+        for (var name in ['error_description', 'error_uri']) {
+          var value = parameters[name];
+
+          if (value != null && value is! String) {
+            throw FormatException(
+                'parameter "$name" was not a string, was "$value"');
+          }
+        }
+
+        var description = parameters['error_description'];
+        var uriString = parameters['error_uri'];
+        var uri = uriString == null ? null : Uri.parse(uriString);
```

Comment:
Consider try/catch around `Uri.parse` and throw a new `FormatException` with a custom message.